File: deck/seed/code/native/kotlin/file/runtime/grip.kt
// Synchronous file handles for the kotlin target, over java.io.RandomAccessFile, which is the JDK type that has a
// cursor. The asynchronous side is java.nio on Dispatchers.IO; this is the same operations with nothing to
// suspend, for the places that cannot wait or should not.
//
// Reached only through the public file/synchronous API.
import java.io.RandomAccessFile

object grip {
  class Grip(val file: RandomAccessFile)

  fun gripOpen(
    path: String,
    read: Boolean,
    write: Boolean,
    create: Boolean,
    append: Boolean,
    clear: Boolean,
  ): Grip {
    val at = java.io.File(path)

    if (create && !at.exists()) {
      at.createNewFile()
    }

    // "rw" is the only writable mode RandomAccessFile has; read-only is "r"
    val file = RandomAccessFile(at, if (write || append || clear) "rw" else "r")

    if (clear && !append) {
      file.setLength(0L)
    }

    if (append) {
      file.seek(file.length())
    }

    return Grip(file)
  }

  fun gripClose(file: Grip) {
    try {
      file.file.close()
    } catch (error: Exception) {
      Unit
    }
  }

  fun gripRead(file: Grip, size: Long): String {
    if (size <= 0L) return ""

    return try {
      val buffer = ByteArray(size.toInt())
      val count = file.file.read(buffer)

      if (count <= 0) "" else String(buffer, 0, count, Charsets.UTF_8)
    } catch (error: Exception) {
      ""
    }
  }

  fun gripWrite(file: Grip, data: String): Long =
    try {
      val bytes = data.toByteArray(Charsets.UTF_8)
      file.file.write(bytes)
      bytes.size.toLong()
    } catch (error: Exception) {
      0L
    }

  fun gripSeek(file: Grip, offset: Long, frame: String) {
    try {
      val at =
        when (frame) {
          "relative" -> file.file.filePointer + offset
          "end" -> file.file.length() - offset
          else -> maxOf(0L, offset)
        }
      file.file.seek(maxOf(0L, at))
    } catch (error: Exception) {
      Unit
    }
  }

  fun gripFlush(file: Grip) {
    try {
      file.file.fd.sync()
    } catch (error: Exception) {
      Unit
    }
  }

  fun gripClear(file: Grip, size: Long) {
    try {
      file.file.setLength(maxOf(0L, size))

      if (file.file.filePointer > size) {
        file.file.seek(maxOf(0L, size))
      }
    } catch (error: Exception) {
      Unit
    }
  }
}

File: deck/seed/code/native/rust/file/runtime/grip.rs
// Synchronous file handles for the rust target, over `std::fs::File`. The asynchronous side is tokio; this is the
// same operations with nothing to await, for the places that cannot wait (reading configuration before a runtime
// is up) or should not (a small read where a task switch costs more than the read).
//
// Reached only through the public file/synchronous API.
mod grip {
    use std::cell::RefCell;
    use std::io::{Read, Seek, Write};
    use std::rc::Rc;

    // Rc so the emitted `synchronous-file` record stays Clone (every emitted struct derives it), RefCell because
    // read, write and seek all take &mut. Two copies of the record are two views of ONE open file, which is what
    // a cursor has to mean.
    pub type Grip = Rc<RefCell<std::fs::File>>;

    pub fn grip_open(
        path: String,
        read: bool,
        write: bool,
        create: bool,
        append: bool,
        clear: bool,
    ) -> Grip {
        let mut options = std::fs::OpenOptions::new();
        options.read(read);

        if append {
            options.append(true);
        } else if write {
            options.write(true);
        }

        options.create(create).truncate(clear && !append);

        Rc::new(RefCell::new(
            options.open(&path).expect("synchronous file open"),
        ))
    }

    // std closes on drop and has no explicit close; flushing is the observable half of one
    pub fn grip_close(file: Grip) {
        let _ = file.borrow_mut().flush();
    }

    pub fn grip_read(file: Grip, size: i64) -> String {
        let mut buffer = vec![0u8; size.max(0) as usize];
        let count = file.borrow_mut().read(&mut buffer).unwrap_or(0);

        String::from_utf8_lossy(&buffer[..count]).to_string()
    }

    pub fn grip_write(file: Grip, data: String) -> i64 {
        let bytes = data.as_bytes();

        match file.borrow_mut().write_all(bytes) {
            Ok(()) => bytes.len() as i64,
            Err(_) => 0,
        }
    }

    pub fn grip_seek(file: Grip, offset: i64, frame: String) {
        let seek = match frame.as_str() {
            "relative" => std::io::SeekFrom::Current(offset),
            "end" => std::io::SeekFrom::End(-offset),
            _ => std::io::SeekFrom::Start(offset.max(0) as u64),
        };
        let _ = file.borrow_mut().seek(seek);
    }

    pub fn grip_flush(file: Grip) {
        let _ = file.borrow_mut().sync_data();
    }

    pub fn grip_clear(file: Grip, size: i64) {
        let _ = file.borrow_mut().set_len(size.max(0) as u64);
    }
}
